File: src/levels/L01-strategy/strategy.rb
class Cashier
  def initialize(payment_methods, barista)
    @payment_methods = payment_methods
    @barista = barista
  end

  # Cashier ya no sabe cómo se cobra: solo a quién delegar.
  # region: Cashier#charge
  def charge(order)
    @payment_methods.fetch(order.payment).charge(order)
    @barista.prepare(order)
  end
  # endregion
end

class CashPayment
  # region: CashPayment#charge
  def charge(order) = CashDrawer.collect(order.total)
  # endregion
end

class CardPayment
  # region: CardPayment#charge
  def charge(order) = CardTerminal.charge(order.total)
  # endregion
end

class VoucherPayment
  # region: VoucherPayment#charge
  def charge(order) = VoucherBook.redeem(order.total)
  # endregion
end

# region: wiring
cashier = Cashier.new(
  { cash: CashPayment.new, card: CardPayment.new, voucher: VoucherPayment.new },
  barista
)
# endregion

File: src/levels/L01-strategy/ts/strategy.ts
interface PaymentMethod {
  charge(order: Order): void
}

class Cashier {
  constructor(
    private paymentMethods: Record<string, PaymentMethod>,
    private barista: Barista,
  ) {}

  // Cashier ya no sabe cómo se cobra: solo a quién delegar.
  // region: Cashier#charge
  charge(order: Order) {
    this.paymentMethods[order.payment].charge(order)
    this.barista.prepare(order)
  }
  // endregion
}

class CashPayment implements PaymentMethod {
  // region: CashPayment#charge
  charge(order: Order) { CashDrawer.collect(order.total) }
  // endregion
}

class CardPayment implements PaymentMethod {
  // region: CardPayment#charge
  charge(order: Order) { CardTerminal.charge(order.total) }
  // endregion
}

class VoucherPayment implements PaymentMethod {
  // region: VoucherPayment#charge
  charge(order: Order) { VoucherBook.redeem(order.total) }
  // endregion
}

// region: wiring
const cashier = new Cashier(
  { cash: new CashPayment(), card: new CardPayment(), voucher: new VoucherPayment() },
  barista,
)
// endregion
